### ISSUE TITLE
Cals from 20211226 are problematic

### ISSUE THREAD

**Issue by schlafly** [desihub/desisurveyops#7]
We had problematic calibration images on 20211226 and have marked all tiles from that night as unsure until we are able to decide if this night is salvageable.  Tiles affected:
2576 3238 3270 3607 6028 6376 6744 8764 11778 22955 23474 24146 25715 41253 41260 41484 41757 41758 41773
(note that some of these are backup 4xxxx series and are less important, but there are a number of dark and bright tiles as well)

Update: Title is incorrect, 20211226, I have decided that we should have included the dashes in the night strings as a matter of policy.  Darn.

**Comment by akremin** [comment]
This should be 20211226 (December not November)

**Comment by julienguy** [comment]
Could be solved with rerunning the preprocessing with option `--bkgsub-for-science` but we would like to have a mechanism in the pipeline to keep a record when special options are used for some observations.

**Comment by schlafly** [comment]
I'm giving up on these tiles.  @marcelo-alvarez , can we mark the following tiles as bad and reprocess so these exposures get EFFTIME = 0, and then I'll clear QA on them?

```
EXPID  TILEID  NIGHT  
int64  int64   int64  
------ ------ --------
115523  25715 20211226
115524  24146 20211226
115525   2576 20211226
115526   6744 20211226
115527   3607 20211226
115528   3270 20211226
115529   3238 20211226
115530   3238 20211226
115531   6376 20211226
115532   6376 20211226
115533   8764 20211226
115534   8764 20211226
115535   6028 20211226
115536   6028 20211226
115537  11778 20211226
115538  22955 20211226
115539  22955 20211226
115540  41484 20211226
115541  41260 20211226
115542  41253 20211226
115543  41757 20211226
115544  41758 20211226
115545  41773 20211226
115574  23474 20211226
```
Thank you!  Note that some of these were observed on other nights, but only their exposures from 20211226 are problematic.

**Comment by marcelo-alvarez** [comment]
@schlafly I will do this after Cori is back up and post here with an update after.

**Comment by marcelo-alvarez** [comment]
I tried to run tsnr for 20211226 with:
```
desi_tsnr_afterburner -o /global/cfs/cdirs/desi/spectro/redux/daily/exposures-daily.fits --prod daily --tile-completeness /global/cfs/cdirs/desi/spectro/redux/daily/tiles-daily.csv --aux /global/cfs/cdirs/desi/survey/observations/SV1/sv1-tiles.fits --gfa-proc-dir /global/cfs/cdirs/desi/survey/GFA/ --compute-skymags --add-badexp --update --nproc 8 --nights 20211226 &>>${HOME}/daily/logfiles/tsnr-20211226.log
```
but got the following error:
```
INFO:desi_tsnr_afterburner:961:one_night: Multiprocessing with 8 procs
INFO:desi_tsnr_afterburner:1018:main: Gathering TSNR tmp tables (over ranks: False, 1)
WARNING:desi_tsnr_afterburner:1033:main: /global/cfs/cdirs/desi/spectro/redux/daily/exposures-daily_tmp.fits not found.
Traceback (most recent call last):
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/code/desispec/daily/bin/desi_tsnr_afterburner", line 1213, in <module>
    sys.exit(main())
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/code/desispec/daily/bin/desi_tsnr_afterburner", line 1046, in main
    tsnr2_expid_table = vstack(tsnr2_expid)
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/conda/lib/python3.9/site-packages/astropy/table/operations.py", line 646, in vstack
    tables = _get_list_of_tables(tables)  # validates input
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/conda/lib/python3.9/site-packages/astropy/table/operations.py", line 53, in _get_list_of_tables
    raise ValueError('no values provided to stack.')
ValueError: no values provided to stack.
```
@akremin have you seen this before?

**Comment by akremin** [comment]
Yes this is a known issue (https://github.com/desihub/desispec/issues/1345),
but we've typically encountered it on nights with no good data to begin
with. This is the first time I am aware of where we flagged the data for a
night as bad and are now trying to overwrite rows in the table with zeros
after flagging.

I looked into the code to solve the error in the past and wasn't able to
immediately solve the issue. We haven't made more of an effort to solve it
because this issue didn't impact very many tiles and in the longterm we
want to rewrite the entire afterburner from scratch. This new situation
sounds like it might elevate the urgency, however.

**Comment by marcelo-alvarez** [comment]
@akremin thanks for the helpful background on the known issue with desi_tsnr_afterburner. Hopefully it can be resolved soon.

Update on the reprocessing:
1. I marked all exposures @schlafly listed above as bad in the exposure table for 20221226 using desi_edit_exposure_table.

2. I purged the corresponding tiles on all nights >= 20221226 using desi_purge_tilenight, which removed the exposures and automatically modified the processing tables in a self-consistent way in preparation for reprocessing. I decided it would be easier to rerun the exposure processing than try to edit the processing table manually and not rerun the exposure processing; @akremin please comment if there is an easier way.

3. The following commands have been run to reprocess tiles after 20221226:
```
desi_run_night -n 20220101 --tiles 41757 --z-submit-types=cumulative --use-specter --append-to-proc-table 
desi_run_night -n 20220316 --tiles 41758 --z-submit-types=cumulative --use-specter --append-to-proc-table 
desi_run_night -n 20220108 --tiles 41484,41757 --z-submit-types=cumulative --use-specter --append-to-proc-table
desi_run_night -n 20220331 --tiles 41758 --z-submit-types=cumulative  --use-specter --append-to-proc-table 
```
with 41757/20220101 failing (exp 116311 not enough stdstars on sp9, not yet flagged), 41484/20220108, 41758/20220316, 41758/20220331 successful, and 41757/20220108 exp 117466 not processed past skysub due to low efftime. 

4. The following commands have yet to be run:
```
desi_run_night -n 20220111 --tiles 41260,41253,41757,41773 --z-submit-types=cumulative --use-specter --append-to-proc-table 
desi_run_night -n 20221121 --tiles 41758 --z-submit-types=cumulative --use-specter --append-to-proc-table 
desi_run_night -n 20220129 --tiles 41484 --z-submit-types=cumulative --use-specter --append-to-proc-table 
desi_run_night -n 20220129 --tiles 41260 --z-submit-types=cumulative --use-specter --append-to-proc-table 
desi_run_night -n 20220227 --tiles 41484 --z-submit-types=cumulative --use-specter --append-to-proc-table
desi_run_night -n 20220131 --tiles 41260 --z-submit-types=cumulative --use-specter --append-to-proc-table 
```
Since the redshifts are cumulative and the cframes were removed by desi_purge_tilenight, a tile on a given night can only be reprocessed once it has been processed on all previous nights; thus the first three commands above are ready to run, but not the last three.

5. In parallel with the above reprocessing in 3 and 4, I attempted to run desi_tsnr_afterburner for nights requiring no further processing, starting with 20221226, but it crashed due to a longstanding issue (see comment above). Consequently I was unable to update exposures-daily.csv, tiles-daily.csv, and exposures-daily.fits together in the usual way for any of the tiles.

6. I made copies of exposures-daily.csv and tiles-daily.csv (consistent with those in daily as of 8:26PM PST Dec 22 2022) that should reflect the correct efftime_spec values for 3238, 3270, 3607, 6028, 6376, 8764, 23474, 24146, 25715. These are located at:
```
/global/cfs/cdirs/desi/users/desi/daily/desisurveyops7/exposures-daily_20221222-826PM.csv
/global/cfs/cdirs/desi/users/desi/daily/desisurveyops7/tiles-daily_20221222-826PM.csv
```
It may be possible to unblock this subset of tiles by copying these files to daily before running desi_tsnr_afterburner for 20221226. However, since I did not manage to do the same for exposures-daily.fits, I did not propagate these changes through to the "official" csv files in daily, in case this would cause unforeseen problems with the offline pipeline. @akremin if you determine it's safe, please feel free to copy the above files and ping @schlafly after. These are tiles (3238, 3270, 3607, 6028, 6376, 8764, 23474, 24146, 25715) for which no further reprocessing is necessary, i.e. lastnight = 20221226.

I should be able to work on this again after about a week. In the meantime please feel free to make progress in my absence.

**Comment by marcelo-alvarez** [comment]
Too late for me to do anything in the next week, but I just realized a simple fix could be to repopulate a single low priority tile by setting laststep back to all on the corresponding exposure on 20211226 and running desi_run_night. Then desi_tsnr_afterburner would not crash and all three files would be updated. The one tile would still be blocking, but all the non-backup tiles would be immediately unblocked since there are no exposures for those after 20211226.

**Comment by schlafly** [comment]
Along these lines, in case you don't know, the tiles with numbers 40000 <= TILEID < 60000 are "backup tiles" which are much less important than other tiles.  If you marked one of those good to resolve this, that would have very little in the way of negative consequences.

**Comment by marcelo-alvarez** [comment]
@schlafly I have marked 115523-115539, 115574 from 20211226 as bad and reprocessed with desi_tsnr_afterburner. The corresponding rows in the exposures-daily
```
rows matching tiles
  25715, 24146, 2576, 6744, 3607, 3270, 3238, 6376, 8764, 6028, 11778, 22955, 23474
in
  /global/cfs/cdirs/desi/spectro/redux/daily/exposures-daily.csv

TILEID EXPTIME  EFFTIME_SPEC  NIGHT   EXPID    EBV
------ -------- ------------ -------- ------ --------
 25715   619.60         0.00 20211226 115523   0.0310

 24146   457.00         0.00 20211226 115524   0.0660

  2576   962.90       311.30 20211222 115159   0.0270
  2576   625.90         0.00 20211226 115525   0.0260

  6744   974.40       629.30 20211222 115157   0.0330
  6744    57.30         0.00 20211222 115158   0.0330
  6744   452.20         0.00 20211226 115526   0.0330

  3607  1134.20         0.00 20211226 115527   0.0430

  3270  1237.70         0.00 20211226 115528   0.0320

  3238  1038.80         0.00 20211226 115529   0.0320
  3238  1066.40         0.00 20211226 115530   0.0320

  6376  1182.90         0.00 20211226 115531   0.0520
  6376   929.30         0.00 20211226 115532   0.0520

  8764   962.90         0.00 20211226 115533   0.0880
  8764   860.80         0.00 20211226 115534   0.0880

  6028  1398.90         0.00 20211226 115535   0.0600
  6028  1437.70         0.00 20211226 115536   0.0600

 11778  1100.70       616.60 20211010 103687   0.0410
 11778   645.90       152.40 20211010 103688   0.0410
 11778  1243.80         0.00 20211226 115537   0.0470

 22955   581.50        96.90 20211008 103456   0.0430
 22955   955.60         0.00 20211226 115538   0.0430
 22955   587.50         0.00 20211226 115539   0.0430

 23474   946.60         0.00 20211226 115574   0.0120
```
and tiles-daily
```
rows matching tiles
  25715, 24146, 2576, 6744, 3607, 3270, 3238, 6376, 8764, 6028, 11778, 22955, 23474
in
  /global/cfs/cdirs/desi/spectro/redux/daily/tiles-daily.csv

TILEID NEXP EXPTIME  EFFTIME_SPEC LASTNIGHT
------ ---- -------- ------------ ---------
 25715    1   619.60         0.00  20211226
 24146    1   457.00         0.00  20211226
  2576    2  1588.80       311.30  20211222
  6744    3  1483.90       629.30  20211222
  3607    1  1134.20         0.00  20211226
  3270    1  1237.70         0.00  20211226
  3238    2  2105.20         0.00  20211226
  6376    2  2112.20         0.00  20211226
  8764    2  1823.70         0.00  20211226
  6028    2  2836.60         0.00  20211226
 11778    3  2990.30       769.00  20211010
 22955    3  2124.60        96.90  20211008
 23474    1   946.60         0.00  20211226
```
files look good to me. If they look good to you, then you should be able to clear QA on tiles 2576, 6744, 3607, 3270, 3238, 6376, 8764, 6028, 11778, 22955, and 23474.

**Comment by marcelo-alvarez** [comment]
All science exposures from 20211226 are currently marked bad except for  exposure 115542 on tile 41253, which was left with LASTSTEP=all, in order to avoid the problem mentioned above, i.e. desi_tsnr_afterburner crashes on nights with no good data to begin with. @akremin has a [PR](https://github.com/desihub/desispec/pull/1950) to fix this under review that should be merged shortly, at which point I will mark 115542 as bad, purge 41253 from 20211226, and re-run desi_tsnr_afterburner.

Finally, note that the other tiles -- i.e. 41484, 41260, 41253, 41757, 41758, 41773, with science exposures on 20211226 -- were observed after 20211226 and have yet to be reprocessed fully. I will post here when that is done and and tsnr has been re-run for all affected nights, at which point we should be able to close the ticket.

**Comment by schlafly** [comment]
I've cleared QA on these tiles.
